src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+//! ## Constraint system for LEM
+//!
+//! This module implements the generation of bellperson constraints for LEM, such
+//! that it can be used with Nova folding to prove evaluations of Lurk expressions.
+//!
+//! Some LEMs may require expensive gadgets, such as Poseidon hashing. So we use
+//! the concept of "slots" to avoid wasting constraints. To explore this idea,
+//! let's use the following LEM as an example:
+//!
+//! ```text
+//! a b c {
+//!     match_tag c {
+//!         Num => {
+//!             let x: Cons = hash2(a, b);
+//!             return (x, x, x);
+//!         },
+//!         Char => {
+//!             let m: Cons = hash2(b, a);
+//!             let n: Cons = hash2(c, a);
+//!             return (m, m, n);
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! On a first impression, one might think that we need to perform three hashing
+//! operations in the circuit when in fact we can get away with only two. That
+//! is so because interpretation can only follow one of the paths:
+//!
+//! * If it goes through `Num`, we need to get one hash right
+//! * If it goes through `Char`, we need to get two hashes right
+//!
+//! Either way, that's at most two hashes that we really care about. So we say
+//! that we need to allocate two slots. The first slot is for the the hash of `x`
+//! or `m` and the second slot if for the hash of `n`. Let's see a sketch of part
```

Comment:
Second slot is for the hash of `n` or a dummy.

Reply:
Fixed